### ISSUE TITLE
Cannot UNEXPORT GPIO pin when universal cape overlay is installed

### ISSUE THREAD

**Issue by jrobparsons** [beagleboard/bb.org-overlays#134]
Why does unexport fail when the universal cape device tree overlay is installed?

**PLATFORM**
```
debian@beaglebone:~$ uname -a
Linux beaglebone 4.9.82-ti-r102 #1 SMP PREEMPT Thu Feb 22 01:16:12 UTC 2018 armv7l GNU/Linux
ddebian@beaglebone:~$ cat /etc/debian_version
9.9
```
**UNIVERSAL CAPE OVERLAY INSTALLED**

/boot/uEnv.txt

```
###Cape Universal Enable
enable_uboot_cape_universal=1
```
Bootup messages:

```
uboot_overlays: loading /lib/firmware/univ-bbb-Exx-00A0.dtbo ...
105226 bytes read in 1067 ms (95.7 KiB/s)
```
GPIO pins (59 pins) are exported:

```
debian@beaglebone:/sys/class/gpio$ ls -l
total 0
--w--w---- 1 root gpio 4096 Aug  6 14:41 export
lrwxrwxrwx 1 root gpio    0 Aug  6 14:41 gpio10 -> ../../devices/platform/ocp/44e07000.gpio/gpiochip0/gpio/gpio10
...
lrwxrwxrwx 1 root gpio    0 Aug  6 14:41 gpio9 -> ../../devices/platform/ocp/44e07000.gpio/gpiochip0/gpio/gpio9
lrwxrwxrwx 1 root gpio    0 Aug  6 14:41 gpiochip0 -> ../../devices/platform/ocp/44e07000.gpio/gpio/gpiochip0
lrwxrwxrwx 1 root gpio    0 Aug  6 14:41 gpiochip32 -> ../../devices/platform/ocp/4804c000.gpio/gpio/gpiochip32
lrwxrwxrwx 1 root gpio    0 Aug  6 14:41 gpiochip64 -> ../../devices/platform/ocp/481ac000.gpio/gpio/gpiochip64
lrwxrwxrwx 1 root gpio    0 Aug  6 14:41 gpiochip96 -> ../../devices/platform/ocp/481ae000.gpio/gpio/gpiochip96
--w--w---- 1 root gpio 4096 Aug  6 14:41 unexport
debian@beaglebone:/sys/class/gpio$
```

Attempt to unexport a pin fails:

```
debian@beaglebone:/sys/class/gpio$ sudo su
[sudo] password for debian:
root@beaglebone:/sys/class/gpio# echo 45 > /sys/class/gpio/unexport
bash: echo: write error: Invalid argument
root@beaglebone:/sys/class/gpio# echo '45' > /sys/class/gpio/unexport
bash: echo: write error: Invalid argument
root@beaglebone:/sys/class/gpio#
```

**UNIVERSAL CAPE OVERLAY NOT INSTALLED**

/boot/uEnv.txt

```
###Cape Universal Enable
#enable_uboot_cape_universal=1
```
Bootup messages:

`uboot_overlays: add [enable_uboot_cape_universal=1] to /boot/uEnv.txt to enable...`

GPIO pins (2 pins) are exported:

```
debian@beaglebone:/sys/class/gpio$ ls -l
total 0
--w--w---- 1 root gpio 4096 Aug  6 16:06 export
lrwxrwxrwx 1 root gpio    0 Aug  6 16:02 gpio12 -> ../../devices/platform/ocp/44e07000.gpio/gpiochip0/gpio/gpio12
lrwxrwxrwx 1 root gpio    0 Aug  6 16:02 gpio13 -> ../../devices/platform/ocp/44e07000.gpio/gpiochip0/gpio/gpio13
lrwxrwxrwx 1 root gpio    0 Aug  6 16:02 gpiochip0 -> ../../devices/platform/ocp/44e07000.gpio/gpio/gpiochip0
lrwxrwxrwx 1 root gpio    0 Aug  6 16:02 gpiochip32 -> ../../devices/platform/ocp/4804c000.gpio/gpio/gpiochip32
lrwxrwxrwx 1 root gpio    0 Aug  6 16:02 gpiochip64 -> ../../devices/platform/ocp/481ac000.gpio/gpio/gpiochip64
lrwxrwxrwx 1 root gpio    0 Aug  6 16:02 gpiochip96 -> ../../devices/platform/ocp/481ae000.gpio/gpio/gpiochip96
--w--w---- 1 root gpio 4096 Aug  6 16:07 unexport
debian@beaglebone:/sys/class/gpio$
```

Attempt to unexport a pin succeeds:

```
debian@beaglebone:/sys/class/gpio$ sudo su
[sudo] password for debian:
root@beaglebone:/sys/class/gpio# ls -l gpio45
ls: cannot access 'gpio45': No such file or directory
root@beaglebone:/sys/class/gpio# echo 45 > /sys/class/gpio/export
root@beaglebone:/sys/class/gpio# ls -l gpio45
lrwxrwxrwx 1 root root 0 Aug  6 16:06 gpio45 -> ../../devices/platform/ocp/4804c000.gpio/gpiochip1/gpio/gpio45
root@beaglebone:/sys/class/gpio# echo 45 > /sys/class/gpio/unexport
root@beaglebone:/sys/class/gpio# ls -l gpio45
ls: cannot access 'gpio45': No such file or directory
root@beaglebone:/sys/class/gpio#
```

**Comment by jrobparsons** [comment]
Here is what I have learned so far...

I see from the univ-bbb-Exx-00A0.dts that the cape-universal node uses the "gpio-of-helper" device driver to configure and export the GPIO pins.

```
	fragment@2 {
		target = <&ocp>;
		__overlay__ {
			cape-universal {
				compatible = "gpio-of-helper";

```

What I do not know, yet, is how to access the "gpio-of-helper" device driver from user space to 'unexport' a pin.

Is there documentation that describes how to use the API supported by the "gpio-of-helper" device driver?

**Comment by jrobparsons** [comment]
I am closing this issue because I now have a solution.

This issue came up while trying to get 2 different Python programs to run that used different interfaces to access their respective GPIO pins. One program uses the 'sysfs' interface. The other program uses the 'libgpiod' interface.

Copied from original issue referenced above...

**1. The sysfs/libgpiod playground**

In order to free up a pin for use by libgpiod_pulseio (via libgpiod interface) and leave pins for use by Adafruit_* modules (via sysfs interface), the universal cape overlay that is loaded during boot-up must be edited, compiled, then copied back into /lib/firmware where available for next boot-up.

In my system, the overlay being loaded is,

univ-bbb-EVA-00A0.dtbo
univ-bbb-EVx-00A0.dtbo
**univ-bbb-Exx-00A0.dtbo**
univ-bbb-xVA-00A0.dtbo
univ-bbb-xVx-00A0.dtbo
univ-bbb-xxx-00A0.dtbo

and is determined by my /boot/uEnv.txt configuration.

I edited the corresponding DTS file in "/opt/source/bb.org-overlays/src/arm/", compiled it using "/opt/source/bb.org-overlays/make src/arm/univ-bbb-Exx-00A0.dtbo", then copied the DTBO file to "/lib/firmware".

I edited the DTS file by disabling all configuration of pin P8_11 (the one used in the dht_simpletst.py example) by following what was done to disable pins P9_19 and P9_20 (the pins supporting the cape I2C EEPROM bus).

**2. Very helpful article!**

Building a device tree overlay for your new PocketCape design
https://beagleboard.org/blog/2018-01-17-building-a-device-tree-overlay-for-your-new-pocketcape-design
